async_processor/amqp_pub_sub.py

```diff
@@ -0,0 +1,116 @@
+            return self._connection
+        self._connection = await connect_robust(self._queue_url)
+        return self._connection
+
+    async def _get_channel(self) -> Channel:
+        if self._channel:
+            return self._channel
+        await self._get_connect()
+        self._channel = await self._connection.channel()
+        return self._channel
+
+    async def _get_queue(self):
+        if self._queue:
+            return self._queue
+        await self._get_channel()
+        self._queue = await self._channel.declare_queue(self._queue_name, durable=True)
+        return self._queue
+
+    @asynccontextmanager
+    async def get_input_message(
+        self,
+    ) -> AsyncIterator[Optional[bytes]]:
+        message = None
+        queue = await self._get_queue()
+        try:
+            message = await queue.get(fail=False, timeout=self._wait_time_seconds)
+        except Exception as ex:
+            raise InputMessageFetchFailure(f"Error fetch input message: {ex}") from ex
+        if not message:
+            yield None
+            return
+        try:
+            yield message.body
+        except Exception as ex:
+            raise InputMessageFetchFailure(
+                f"Error decoding input message body: {ex}"
+            ) from ex
+        finally:
+            try:
+                await message.ack()
+            except Exception as ex:
+                raise InputFetchAckFailure(
+                    f"Error publishing input message: {ex}"
+                ) from ex
+
+    async def publish_input_message(
+        self, serialized_input_message: bytes, request_id: str
+    ):
+        channel = await self._get_channel()
+        await channel.default_exchange.publish(
+            Message(body=serialized_input_message), routing_key=self._queue_name
+        )
+
+
+class AMQPOutput(Output):
+    def __init__(self, config: AMQPOutputConfig):
+        self._queue_url = config.queue_url
+        self._queue_name = config.queue_name
+        self._connection = None
+        self._channel = None
+        self._queue = None
+
+    async def _get_connect(self):
+        if self._connection:
+            return self._connection
+        self._connection = await connect_robust(self._queue_url)
+        return self._connection
+
+    async def _get_channel(self) -> Channel:
+        if self._channel:
+            return self._channel
+        await self._get_connect()
+        self._channel = await self._connection.channel()
+        return self._channel
+
+    async def _get_queue(self):
+        if self._queue:
+            return self._queue
+        await self._get_channel()
+        self._queue = await self._channel.declare_queue(self._queue_name, durable=True)
+        return self._queue
+
+    async def publish_output_message(
+        self, serialized_output_message: bytes, request_id: Optional[str]
+    ):
+        channel = await self._get_channel()
+        await self._get_queue()
```

Comment:
This seems odd to me. Why do we need to call it here?

Reply:
Yes, I added this to create the queue
if we don't add this, the data will be skipped

Reply:
Ideally, the queue should be created by the Ops/Developer.

Reply:
When we start, we should validate that the queue exists. Error out otherwise.
https://github.com/truefoundry/async_service/blob/c5d92379a631ebba71353e5a275304dc4285175a/async_processor/nats_pub_sub.py#L50

Reply:
I have added an example above for NATS.